File: 2019/1.go
package main

import (
        "fmt"
        "io/ioutil"
        "strconv"
        "strings"
)

func calc_fuel(mass int) int {
        var fuel int 
        for {
                new_fuel := (mass / 3) - 2 
                if new_fuel < 0 { 
                        break
                }
                fuel += new_fuel
                mass = new_fuel                                                                                                                                                                                                                                                                                                                  
        }
        return fuel
}

func check(e error) {
        if e != nil {
                panic(e)
        }
}

func main() {
        dat, err := ioutil.ReadFile("1_input.txt")
        check(err)
        mass_list := strings.Split(string(dat), "\n")
        var total_fuel int = 0 
        for i := 0; i < len(mass_list); i++ {
                fmt.Printf("\"%s\"\n", mass_list[i])
                num, err := strconv.Atoi(mass_list[i])
                if err != nil {
                        continue
                }
                total_fuel += calc_fuel(num)
        }
        fmt.Printf("Total fuel: %d\n", total_fuel)
}

File: 2019/1.rs
use std::fs;

fn calc_fuel(mass: i32) -> i32 {
    let mut fuel = (mass / 3) - 2;
    if fuel > 0 {
        let new_fuel = calc_fuel(fuel);
        if new_fuel > 0 {
            fuel += new_fuel;
        }
    }
    return fuel
}

fn parse_input(filename: &str) {

    println!("In file {}", filename);

    let contents = fs::read_to_string(filename)
        .expect("Something went wrong reading the file");

    let mass_list = contents.trim().split("\n");
    let mut total_fuel = 0;
    for mass in mass_list {
        let mass_int = mass.parse::<i32>().unwrap();
        total_fuel += calc_fuel(mass_int)
    }   
    println!("{}", total_fuel)
}

fn main() {
    parse_input("1_input.txt")
}
